plugin/plugin.go

```diff
@@ -0,0 +1,209 @@
+	IncidentSummary   string `envconfig:"PLUGIN_INCIDENT_SUMMARY"`
+	IncidentSource    string `envconfig:"PLUGIN_INCIDENT_SOURCE"`
+	IncidentSeverity  string `envconfig:"PLUGIN_INCIDENT_SEVERITY"`
+	DedupKey          string `envconfig:"PLUGIN_DEDUP_KEY"`
+	CreateChangeEvent bool   `envconfig:"PLUGIN_CREATE_CHANGE_EVENT"`
+	Resolve           bool   `envconfig:"PLUGIN_RESOLVE"`
+	JobStatus         string `envconfig:"PLUGIN_JOB_STATUS"`
+	CustomDetailsStr  string `envconfig:"PLUGIN_CUSTOM_DETAILS"` // Intermediate string to receive JSON
+	CustomDetails     map[string]interface{}
+}
+
+// PagerDutyClient defines the methods used from the PagerDuty API.
+type PagerDutyClient interface {
+	ManageEventWithContext(ctx context.Context, event *pagerduty.V2Event) (*pagerduty.V2EventResponse, error)
+	CreateChangeEventWithContext(ctx context.Context, event pagerduty.ChangeEvent) (*pagerduty.ChangeEventResponse, error)
+}
+
+// validateSeverity validates PagerDuty's allowed severity values.
+func validateSeverity(severity string) error {
+	switch severity {
+	case SeverityCritical, SeverityError, SeverityWarning, SeverityInfo, SeverityUnknown:
+		return nil
+	default:
+		return errors.New("invalid severity value; allowed values are 'critical', 'error', 'warning', 'info', 'unknown'")
+	}
+}
+
+// Exec executes the plugin.
+func Exec(ctx context.Context, client PagerDutyClient, args Args) error {
+	logger := logrus.WithFields(logrus.Fields{
+		"PLUGIN_ROUTING_KEY":         args.RoutingKey,
+		"PLUGIN_INCIDENT_SUMMARY":    args.IncidentSummary,
+		"PLUGIN_INCIDENT_SOURCE":     args.IncidentSource,
+		"PLUGIN_INCIDENT_SEVERITY":   args.IncidentSeverity,
+		"PLUGIN_CREATE_CHANGE_EVENT": args.CreateChangeEvent,
+		"PLUGIN_JOB_STATUS":          args.JobStatus,
+	})
+
+	logger.Info("Starting plugin execution")
+
+	// Validate required fields
+	if args.RoutingKey == "" {
+		return errors.New("missing required parameter: routingKey")
+	}
+	if args.IncidentSummary == "" {
+		return errors.New("missing required parameter: incidentSummary")
+	}
+	if args.IncidentSource == "" {
+		return errors.New("missing required parameter: incidentSource")
+	}
+	if !args.CreateChangeEvent {
+		if args.DedupKey == "" {
+			return errors.New("missing required parameter: dedupKey when not creating a change event")
+		}
+		if args.JobStatus == "" {
+			return errors.New("missing required parameter: jobStatus when not creating a change event")
+		}
+	}
+
+	// Validate severity value if not creating a change event
+	if !args.CreateChangeEvent {
+		if err := validateSeverity(args.IncidentSeverity); err != nil {
+			return err
+		}
+	}
+
+	if args.JobStatus == "" {
+		logger.Warn("Job status is empty, no action will be taken")
+	}
+
+	if args.CreateChangeEvent {
+		logger.Info("Creating change event")
+		if err := createChangeEvent(ctx, client, args); err != nil {
+			logger.WithError(err).Error("Failed to create change event")
+			return errors.New("failed to create change event")
```

Comment:
this loks more generic, could you please include the actual error for better debugging